### PR TITLE
[LiveIns] Improve recomputeLiveIns()

### PR DESCRIPTION
**Pull request by redstar**
Some small changes to recomputeLiveIns() to improve performance:

- Instead of copying the list of old live-ins, and then clearing
  them, a new method swaps the list for an empty one.
- getLiveIns() now returns a constant reference to the list

As result, the list-data is never copied. Depending on the implementation
details of the vector container, it can also save calls to allocate
and deallocate memory.
I see a small improvement on CTMark with these changes.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/LivePhysRegs.h [hunk 1]
```diff
@@ -197,16 +197,16 @@ void computeAndAddLiveIns(LivePhysRegs &LiveRegs,
 
 /// Convenience function for recomputing live-in's for a MBB. Returns true if
 /// any changes were made.
-static inline bool recomputeLiveIns(MachineBasicBlock &MBB) {
+inline bool recomputeLiveIns(MachineBasicBlock &MBB) {
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/LivePhysRegs.h:200]: This needs the extern definition somewhere. Maybe just move the body out of line?

Comment (redstar) [reply llvm/include/llvm/CodeGen/LivePhysRegs.h:200]: Thanks - the `static` is there for ages, I forgot to remove it.

File: llvm/include/llvm/CodeGen/LivePhysRegs.h [hunk 2]
```diff
@@ -197,16 +197,16 @@ void computeAndAddLiveIns(LivePhysRegs &LiveRegs,
...
+  std::vector<MachineBasicBlock::RegisterMaskPair> OldLiveIns;
 
-  MBB.clearLiveIns();
+  MBB.clearLiveIns(OldLiveIns);
   computeAndAddLiveIns(LPR, MBB);
   MBB.sortUniqueLiveIns();
 
-  auto newLiveIns = MBB.getLiveIns();
-  return oldLiveIns != newLiveIns;
+  auto NewLiveIns = MBB.getLiveIns();
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/LivePhysRegs.h:208]: const & no auto?

File: llvm/include/llvm/CodeGen/MachineBasicBlock.h [hunk 3]
```diff
@@ -477,7 +481,7 @@ class MachineBasicBlock
   /// Remove entry from the livein set and return iterator to the next.
   livein_iterator removeLiveIn(livein_iterator I);
 
-  std::vector<RegisterMaskPair> getLiveIns() const { return LiveIns; }
+  const std::vector<RegisterMaskPair> getLiveIns() const { return LiveIns; }
```
Comment (nikic) [inline llvm/include/llvm/CodeGen/MachineBasicBlock.h:484]: Was this supposed to be?
```suggestion
  const std::vector<RegisterMaskPair> &getLiveIns() const { return LiveIns; }
```

File: llvm/include/llvm/CodeGen/LivePhysRegs.h (hunk 1 above)
Comment (arsenm) [reply llvm/include/llvm/CodeGen/LivePhysRegs.h:200]: Yes, I mean you need to put it back or ensure this is defined somewhere

File: llvm/include/llvm/CodeGen/LivePhysRegs.h (hunk 2 above)
Comment (redstar) [reply llvm/include/llvm/CodeGen/LivePhysRegs.h]: Changed.

File: llvm/include/llvm/CodeGen/LivePhysRegs.h (hunk 1 above)
Comment (redstar) [reply llvm/include/llvm/CodeGen/LivePhysRegs.h]: Fixed.

**Review by arsenm — APPROVED** [review]

**Pull request merged**